lib/Logger/LogContext.h

```diff
@@ -0,0 +1,559 @@
+  auto with(Value v) && {
+    auto vals =
+        std::tuple_cat(std::move(_vals), std::make_tuple(std::forward<Value>(v)));
+    return ValueBuilder<decltype(vals), Keys..., Key>(std::move(vals));
+  }
+  std::shared_ptr<Values> share() && {
+    return std::make_shared<ValuesImpl<Vals, Keys...>>(std::move(_vals));
+  }
+
+ private:
+  friend class LogContext;
+  ValueBuilder(Vals vals) : _vals(std::move(vals)) {}
+  Vals _vals;
+};
+
+struct LogContext::Values {
+  virtual ~Values() = default;
+  virtual void visit(Visitor const&) const = 0;
+};
+
+template <class Vals, const char... Keys[]>
+struct LogContext::ValuesImpl final : Values {
+  explicit ValuesImpl(Vals vals) : _values(std::move(vals)) {}
+  void visit(Visitor const& visitor) const override {
+    doVisit<0, Keys...>(visitor);
+  }
+  ValuesImpl const* operator->() const { return this; }
+
+ private:
+  template <std::size_t Idx, const char K[], const char... Ks[]>
+  void doVisit(Visitor const& visitor) const {
+    visitor.visit(K, toVisitorType(std::get<Idx>(_values)));
+    if constexpr (Idx + 1 < std::tuple_size<Vals>{}()) {
+      doVisit<Idx + 1, Ks...>(visitor);
+    }
+  }
+  
+  template <class T>
+  static auto toVisitorType(T&& v) {
+    using TT = std::remove_reference_t<T>;
+    if constexpr (std::is_floating_point_v<TT>) {
+      return static_cast<double>(v);
+    } else if constexpr (std::is_unsigned_v<TT>) {
+      return static_cast<std::uint64_t>(v);
+    } else if constexpr (std::is_signed_v<TT>) {
+      return static_cast<std::int64_t>(v);
+    } else if constexpr (std::is_constructible_v<std::string_view, T>) {
+      return std::string_view(std::forward<T>(v));
+    } else if constexpr (std::is_constructible_v<std::string, T>) {
+      return std::string(std::forward<T>(v));
+    } else {
+      std::ostringstream ss;
+      ss << v;
+      return std::move(ss).str();
+    }
+  }
+
+  Vals _values;
+};
+
+struct LogContext::Entry {
+  virtual ~Entry() {
+    TRI_ASSERT(_refCount.load() <= 1);
+  }
+  virtual void visit(Visitor const&) const = 0;
+  
+ protected:
+  virtual void release(EntryCache& cache) noexcept {
+    this->~Entry();
+    ::operator delete(this);
+  }
+
+ private:
+  void incRefCnt() noexcept {
+    TRI_ASSERT(_refCount.load() > 0);
+    _refCount.fetch_add(1, std::memory_order_acquire);
+  }
+  [[nodiscard]] std::size_t decRefCnt() noexcept {
+    TRI_ASSERT(_refCount.load() > 0);
+    return _refCount.fetch_sub(1, std::memory_order_relaxed);
+  }
+
+  friend class LogContext;
+  std::atomic<std::size_t> _refCount{0};
+  Entry* _prev;
+};
+
+template <class Vals>
+struct LogContext::EntryImpl final : Entry {
+  explicit EntryImpl(Vals&& v) : _values(std::move(v)) {}
+  void visit(Visitor const& visitor) const override { _values->visit(visitor); }
+  void release(EntryCache& cache) noexcept override;
+
+ private:
+  Vals _values;
+};
+
+struct LogContext::EntryCache {
+  struct CachedEntryAlloc {
```

Comment:
A `CachedEntryAlloc` instance can be created without specifying a value for `next`, so it will have an undefined value. What about making construction safer by requiring to set the `next` value in the constructor?
```
struct CachedEntryAlloc {
  explicit CachedEntryAlloc(CachedEntryAlloc* n) noexcept { next = n; }
  CachedEntryAlloc* next;
};